apps/aecore/src/aec_db_gc.erl

```diff
@@ -0,0 +1,249 @@
+            ets:delete(Hashes),
+            {next_state, idle, Data#data{height = undefined, hashes = undefined},
+             {reply, From, {ok, N}}};
+        micro ->
+            {keep_state, Data}
+    end;
+handle_event({call, From}, maybe_garbage_collect, _, Data) ->
+    {keep_state, Data, {reply, From, nop}};
+
+handle_event(_, _, _, Data) ->
+    {keep_state, Data}.
+
+
+terminate(_Reason, _State, _Data) -> void.
+
+code_change(_, State, Data, _) -> {ok, State, Data}.
+
+callback_mode() -> handle_event_function.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+%% From - To (inclusive)
+collect_reachable_hashes(FromHeight, ToHeight) when FromHeight < ToHeight ->
+    {ok, Hashes} = collect_reachable_hashes_fullscan(FromHeight),     % table is created here
+    {ok, Hashes} = range_collect_reachable_hashes(FromHeight, ToHeight, Hashes), % and reused
+    {ok, Hashes}.
+
+collect_reachable_hashes_fullscan(Height) ->
+    Tab = ets:new(gc_reachable_hashes, [public]),
+    MPT = get_mpt(Height),
+    ?LOG("scanning accounts full tree with root hash ~w at height = ~p~n",
+         [aeu_mp_trees:root_hash(MPT), Height]),
+    {ok, aeu_mp_trees:visit_reachable_hashes(MPT, Tab, fun store_hash/3)}.
+
+%% assumes Height - 1, Height - 2, ... down to Height - History
+%% are in Hashes from previous runs
+collect_reachable_hashes_delta(Height, Hashes) ->
+    MPT = get_mpt(Height),
+    ?LOG("scanning accounts changes with root hash ~w at height = ~p~n",
+         [aeu_mp_trees:root_hash(MPT), Height]),
+    {ok, aeu_mp_trees:visit_reachable_hashes(MPT, Hashes, fun store_unseen_hash/3)}.
+
+range_collect_reachable_hashes(ToHeight, #data{height = LastHeight, hashes = Hashes}) ->
+    range_collect_reachable_hashes(LastHeight, ToHeight, Hashes).
+range_collect_reachable_hashes(LastHeight, ToHeight, Hashes) ->
+    [collect_reachable_hashes_delta(H, Hashes) || H <- lists:seq(LastHeight + 1, ToHeight)],
+    {ok, Hashes}.
+
+
+%% the actual GC: clear whole table + insert reachable nodes
+swap_nodes(Hashes) ->
+    NodesCount = ets:info(Hashes, size),
+    {ClearTime, {atomic, ok}} = ?TIMED(mnesia:clear_table(aec_account_state)),
+    ?LOG("clearing accounts table took ~p seconds", [ClearTime / 1000000]),
+    ?LOG("writing ~p reachable account state nodes...", [NodesCount]),
+    {WriteTime, ok} =
+        ?TIMED(aec_db:ensure_transaction(
+                 fun () ->
+                         ets:foldl(
+                           fun ({Hash, Node}, ok) ->
+                                   aec_db:write_accounts_node(Hash, Node)
+                           end, ok, Hashes)
+                 end, [], sync_transaction)),
+    ?LOG("writing reachable account state nodes took ~p seconds", [WriteTime / 1000000]),
+    {atomic, DBCount} = mnesia:transaction(fun () -> length(mnesia:all_keys(aec_account_state)) end),
```

Comment:
Ouch! Does this have to be done inside a transaction?
`length(mnesia:dirty_select(aec_account_state, [{'_',[],[1]}]))` ought to be much cheaper.

Reply:
Shouldn't be needed in transaction, but serves more as a sanity check at least in this phase. I expect that the whole counting will be removed later.